packages/nimble-components/src/wafer-map/modules/rendering.ts

```diff
@@ -5,41 +5,72 @@ import type { DataManager } from './data-manager';
+    private readonly dimensions: Dimensions;
+    private readonly labelFontSize: number;
 
     public constructor(waferData: DataManager, canvas: HTMLCanvasElement) {
-        this.waferData = waferData;
         this.context = canvas.getContext('2d')!;
+        this.dies = waferData.diesRenderInfo;
+        this.dimensions = waferData.dieDimensions;
+        this.labelFontSize = waferData.labelsFontSize;
     }
 
-    public drawWafer(): void {
-        const dies: DieRenderInfo[] = this.waferData.diesRenderInfo;
-        const dimensions: Dimensions = this.waferData.dieDimensions;
+    public drawWafer(transform?: number): void {
+        this.renderDies(transform);
+        this.renderText(transform);
+    }
 
-        for (const die of dies) {
-            this.context.fillStyle = die.fillStyle;
-            this.context?.fillRect(
-                die.x,
-                die.y,
-                dimensions.width,
-                dimensions.height
-            );
+    public clearCanvas(width: number, height: number): void {
+        this.context.clearRect(0, 0, width, height);
+    }
 
-            this.context.font = this.waferData.labelsFontSize.toString();
-            this.context.fillStyle = '#ffffff';
-            this.context.textAlign = 'center';
-            const aproxTextHeight = this.context.measureText('M');
+    private renderDies(transform?: number): void {
+        this.dieSize = this.dimensions.width * this.dimensions.height * (transform || 1);
+        this.dies.sort((a, b) => a.fillStyle.localeCompare(b.fillStyle));
+        let prev!: DieRenderInfo;
 
-            this.context.fillText(
-                die.text,
-                die.x + dimensions.width / 2,
-                die.y + dimensions.height / 2 + aproxTextHeight.width / 2
+        for (const die of this.dies) {
+            if (!prev) {
+                this.context.fillStyle = die.fillStyle;
+            }
+            if (prev && die.fillStyle !== prev.fillStyle && die.fillStyle) {
+                this.context.fillStyle = die.fillStyle;
+            }
+            this.context?.beginPath();
+            this.context?.rect(
+                die.x,
+                die.y,
+                this.dimensions.width,
+                this.dimensions.height
             );
+            this.context?.fill();
+            prev = die;
         }
     }
 
-    public clearCanvas(width: number, height: number): void {
-        this.context.clearRect(0, 0, width, height);
+    private renderText(transform?: number): void {
+        this.dieSize = this.dimensions.width * this.dimensions.height * (transform || 1);
+        const fontsize = this.labelFontSize;
+        this.context.font = `${fontsize.toString()}px sans-serif`;
+        this.context.fillStyle = '#ffffff';
+        this.context.textAlign = 'center';
+        this.context.lineCap = 'butt';
+        const aproxTextHeight = this.context.measureText('M');
+
+        if (this.dieSize >= 50) {

```

Comment:
It would be good to add a test case to cover the case where the dies aren't big enough to render the text. You could do this in storybook with a small wafer or a wafer with lots of dies.

Reply:
I went with the first option as the data generator generates random data, thus, we can't use it for matrix tests. The new wafer is a 70x70 and it does not show the  text unless it's zoomed.